// addPage/index.js
#! /usr/bin/env node --harmony

const Path = require( 'path' );
const Fs = require( 'fs' );
const _ = require( 'lodash' );
const yargs = require( 'yargs' );

// ------------------------------------------
// cli

let argv = yargs
	.usage( '$0 <cmd> [args]' )
	.option( 'name', {
		alias: 'n',
		demand: true,
		describe: 'Page Name: Capital-First-Dash-Format'
	} )
	.option( 'delete', {
		alias: 'd',
		boolean: true,
		default: false,
		describe: 'delete a page from your site'
	} )
	.help( 'help' )
	.argv;

let name = argv.name;

// ------------------------------------------
// START

if ( !argv.delete ) {
	addPage( name );
} else {
	removePage( name );
}

// ------------------------------------------

function addPage( name ) {
	makeJS( name );
	makePUG( name );
	makeSASS( name );
	addToMainMenu( name );
	addToJSIndex( name );
	addToPugIndex( name );
	addToSassIndex( name );
}

// ------------------------------------------

function removePage( name ) {
	deleteJS( name );
	deletePUG( name );
	deleteSASS( name );
	removeFromMainMenu( name );
	removeFromJSIndex( name );
	removeFromPugIndex( name );
	removeFromSassIndex( name );
}

// ------------------------------------------

function makeJS( name ) {
	let template = Fs.readFileSync( Path.join( __dirname, 'templates', 'page.js.ust' ), 'utf-8' );
	let rendered = _.template( template )( {
		name
	} );
	Fs.writeFileSync( Path.resolve( __dirname, '../..', 'src/js/views/pages', `${name}-Page.js` ), rendered );
	return rendered;
}

// ------------------------------------------

function makePUG( name ) {
	let template = Fs.readFileSync( Path.join( __dirname, 'templates', 'page.pug.ust' ), 'utf-8' );
	let rendered = _.template( template )( {
		name
	} );
	Fs.writeFileSync( Path.resolve( __dirname, '../..', 'src/pug/static/pages', `_${name}-Page.pug` ), rendered );
	return rendered;
}

// ------------------------------------------

function makeSASS( name ) {
	let template = Fs.readFileSync( Path.join( __dirname, 'templates', 'page.sass.ust' ), 'utf-8' );
	let rendered = _.template( template )( {
		name
	} );
	Fs.writeFileSync( Path.resolve( __dirname, '../..', 'src/sass/pages', `_${name}-Page.sass` ), rendered );
	return rendered;
}

// ------------------------------------------

function addToMainMenu() {
	let srcPath = Path.resolve( __dirname, '../..', 'src/pug/static/ui/_main-menu.pug' );
	let fileContents = Fs.readFileSync( srcPath, 'utf-8' );
	fileContents += `\ta.closer(href="#/${name}") ${name} Page\n`;
	Fs.writeFileSync( srcPath, fileContents );
	return fileContents;
}

// ------------------------------------------

function addToJSIndex( name ) {
	let importsMarker = '// imports\n';
	let exportsMarker = '// exports\n';
	let varName = _.upperFirst( _.camelCase( name ) );
	let srcPath = Path.resolve( __dirname, '../..', 'src/js/views/pages/index.js' );
	let fileContents = Fs.readFileSync( srcPath, 'utf-8' );
	let parts = fileContents.split( importsMarker );
	parts[ 1 ] = `var ${varName} = require( './${name}-Page' );\n` + parts[ 1 ];
	fileContents = parts.join( importsMarker );
	parts = fileContents.split( exportsMarker );
	parts[ 1 ] = `\tnew ${varName}(),\n` + parts[ 1 ];
	fileContents = parts.join( exportsMarker );

	Fs.writeFileSync( srcPath, fileContents );

	return fileContents;

}

// ------------------------------------------

function addToPugIndex( name ) {
	let importsMarker = '// imports\n';
	let srcPath = Path.resolve( __dirname, '../..', 'src/pug/static/index.pug' );
	let fileContents = Fs.readFileSync( srcPath, 'utf-8' );
	let parts = fileContents.split( importsMarker );
	// include pages/_control-panel-page
	parts[ 1 ] = `\t\t\t\tinclude pages/_${name}-Page\n` + parts[ 1 ];
	fileContents = parts.join( importsMarker );

	Fs.writeFileSync( srcPath, fileContents );

	return fileContents;

}

// ------------------------------------------

function addToSassIndex() {
	let srcPath = Path.resolve( __dirname, '../..', 'src/sass/pages/_index.sass' );
	let fileContents = Fs.readFileSync( srcPath, 'utf-8' );
	fileContents += `@import "${name}-Page"\n`;
	Fs.writeFileSync( srcPath, fileContents );
	return fileContents;
}

// ------------------------------------------
// REMOVE PAGE FUNCTIONS
// ------------------------------------------

function deleteJS( name ) {
	Fs.unlinkSync( Path.resolve( __dirname, '../..', 'src/js/views/pages', `${name}-Page.js` ) );
	return name;
}

// ------------------------------------------

function deletePUG( name ) {
	Fs.unlinkSync( Path.resolve( __dirname, '../..', 'src/pug/static/pages', `_${name}-Page.pug` ) );
	return name;
}

// ------------------------------------------

function deleteSASS( name ) {
	Fs.unlinkSync( Path.resolve( __dirname, '../..', 'src/sass/pages', `_${name}-Page.sass` ) );
	return name;
}

// ------------------------------------------

function removeFromMainMenu() {
	let srcPath = Path.resolve( __dirname, '../..', 'src/pug/static/ui/_main-menu.pug' );
	let fileContents = Fs.readFileSync( srcPath, 'utf-8' );

	// find line with name
	let lines = fileContents.split( '\n' );
	// filter lines w/ the name
	fileContents = _.filter( lines, ( text ) => ( text.search( name ) === -1 ) )
		.join( '\n' );
	Fs.writeFileSync( srcPath, fileContents );
	return fileContents;
}

// ------------------------------------------

function removeFromJSIndex( name ) {
	let varName = _.upperFirst( _.camelCase( name ) );
	let srcPath = Path.resolve( __dirname, '../..', 'src/js/views/pages/index.js' );
	let fileContents = Fs.readFileSync( srcPath, 'utf-8' );

	// find line with name
	let lines = fileContents.split( '\n' );
	// filter lines w/ the name
	fileContents = _.filter( lines, ( text ) => ( text.search( varName ) === -1 ) )
		.join( '\n' );

	Fs.writeFileSync( srcPath, fileContents );

	return fileContents;

}

// ------------------------------------------

function removeFromPugIndex( name ) {
	let srcPath = Path.resolve( __dirname, '../..', 'src/pug/static/index.pug' );
	let fileContents = Fs.readFileSync( srcPath, 'utf-8' );

	// find line with name
	let lines = fileContents.split( '\n' );
	// filter lines w/ the name
	fileContents = _.filter( lines, ( text ) => ( text.search( name ) === -1 ) )
		.join( '\n' );

	Fs.writeFileSync( srcPath, fileContents );

	return fileContents;

}

// ------------------------------------------

function removeFromSassIndex() {
	let srcPath = Path.resolve( __dirname, '../..', 'src/sass/pages/_index.sass' );
	let fileContents = Fs.readFileSync( srcPath, 'utf-8' );

	// find line with name
	let lines = fileContents.split( '\n' );
	// filter lines w/ the name
	fileContents = _.filter( lines, ( text ) => ( text.search( name ) === -1 ) )
		.join( '\n' );

	Fs.writeFileSync( srcPath, fileContents );
	return fileContents;
}
